etcd.py

```diff
@@ -62,59 +70,93 @@ def set_info_from_ec2_instance(self, instance):
+
         tags = tags_to_dict(instance.tags)
-        self.cluster_token = tags['aws:cloudformation:stack-name']
+        self.cloudformation_stack = tags[self.CF_TAG]
         self.autoscaling_group = tags[self.AG_TAG]
 
     @staticmethod
     def get_addr_from_urls(urls):
         for url in urls:
             url = urlparse(url)
             if url and url.netloc:
-                # TODO: check that hostname contains ip
                 return url.hostname
         return None
 
+    def addr_matches(self, peer_urls):
+        t = '{0}:' + str(self.peer_port)
+        for url in peer_urls:
+            url = urlparse(url)
+            if url and url.netloc and url.netloc in (t.format(self.private_ip_address),
+                                                     t.format(self.public_ip_address),
+                                                     t.format(self.private_dns_name),
+                                                     t.format(self.public_dns_name)):
+                return True
+        return False
+
     def set_info_from_etcd(self, info):
         # by convention member.name == instance.id
         if self.instance_id and info['name'] and self.instance_id != info['name']:
             return
 
         addr = self.get_addr_from_urls(info['peerURLs'])
         # when you add new member it doesn't have name, but we can match it by peer_addr
-        if self.addr and (not addr or self.addr != addr):
+        if not addr:
             return
+        elif re.match(r'^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$', addr):
+            if (self.private_ip_address or self.public_ip_address) and \
+                    addr not in (self.private_ip_address, self.public_ip_address):
+                return
+            self._addr = addr
+        else:
+            if (self.private_dns_name or self.public_dns_name) and \
+                    addr not in (self.private_dns_name, self.public_dns_name):
+                return
+            self._dns = addr
 
         self.id = info['id']
         self.name = info['name']
         self.client_urls = info['clientURLs']
         self.peer_urls = info['peerURLs']
-        self.addr = addr
 
     @staticmethod
     def generate_url(addr, port):
         return 'http://{}:{}'.format(addr, port)
 
     def get_client_url(self, endpoint=''):
-        url = self.generate_url(self.addr, self.client_port)
+        url = self.generate_url(self.advertise_addr, self.client_port)
         if endpoint:
             url += self.API_VERSION + endpoint
         return url
 
     @property
+    def addr(self):
+        return EtcdCluster.is_multiregion() and self.public_ip_address or self.private_ip_address
+
+    @property
+    def dns(self):
+        return EtcdCluster.is_multiregion() and self.public_dns_name or self.private_dns_name
+
+    @property
+    def advertise_addr(self):
+        return EtcdCluster.is_multiregion() and self.public_dns_name or self.private_ip_address
```

Comment:
Depending on type of deployment it will advertise wither private ip or public dns.